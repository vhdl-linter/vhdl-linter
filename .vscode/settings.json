{
  "typescript.tsdk": "node_modules/typescript/lib",
  "eslint.lintTask.enable": true,
  "eslint.lintTask.options": "--ext .ts .",
  "eslint.format.enable": true,
  "cSpell.words": [
    "actuals",
    "clientrc",
    "downto",
    "Inoutput",
    "Mentionables",
    "nand",
    "Onig",
    "OSVVM",
    "quickfix",
    "Referencable",
    "setq",
    "subarchitecture",
    "Textmate",
    "ulogic",
    "uninstantiated",
    "Uniquify",
    "Verilog",
    "vhdl",
    "Vhdl",
    "vmode",
    "vprop",
    "vunit",
    "Whitespaces",
    "xnor"
  ]
}