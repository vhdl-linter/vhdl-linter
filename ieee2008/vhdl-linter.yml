rules:
  type-resolved: false
  unit: false