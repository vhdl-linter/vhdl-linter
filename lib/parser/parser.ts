/* eslint-disable @typescript-eslint/no-explicit-any */
import { ArchitectureParser } from './architecture-parser';
import { ContextParser } from './context-parser';
import { ContextReferenceParser } from './context-reference-parser';
import { EntityParser } from './entity-parser';
import { MagicCommentType, OFile, OI, OIRange, OMagicCommentDisable, OMagicCommentParameter, ParserError, OConfiguration, OUseClause, OLibrary, OPackageInstantiation } from './objects';
import { PackageParser } from './package-parser';
import { ParserBase } from './parser-base';
import { UseClauseParser } from './use-clause-parser';
import { CancelationObject } from '../language-server';
import { Lexer, OLexerToken, TokenType } from '../lexer';
import { PackageInstantiationParser } from './package-instantiation-parser';

export class ParserPosition {
  public lexerTokens: OLexerToken[];
  public file: OFile;
  public num = 0;
  public get i() {
    if (this.num >= this.lexerTokens.length) {
      throw new ParserError(`I out of range`, this.lexerTokens[this.lexerTokens.length - 1].range);
    }
    return this.lexerTokens[this.num].range.start.i;
  }
  public isLast() {
    return this.num === this.lexerTokens.length - 1;
  }
  public isValid() {
    return this.num >= 0 && this.num < this.lexerTokens.length;
  }
  public getRangeToEndLine() {
    return this.lexerTokens[this.num].range.copyExtendEndOfLine();
  }

}
export class Parser extends ParserBase {
  pos: ParserPosition;
  private originalText: string;
  public lexerTokens: OLexerToken[] = [];
  text: string;
  file: OFile;
  constructor(text: string, filePath: string, public onlyEntity: boolean = false, public cancelationObject: CancelationObject) {
    super(new ParserPosition(), filePath);
    this.originalText = text;
    this.text = text;
    this.file = new OFile(this.text, this.filePath, this.originalText);

    const lexer = new Lexer(this.originalText, this.file);
    this.lexerTokens = lexer.lex();
    this.pos.lexerTokens = this.lexerTokens;
    this.pos.file = this.file;
  }
  parse(): OFile {

    let disabledRangeStart = undefined;
    const ignoreRegex: RegExp[] = [];
    for (const [lineNumber, line] of this.originalText.split('\n').entries()) {
      const match = /(--\s*vhdl-linter)(.*)/.exec(line); // vhdl-linter-disable-next-line //vhdl-linter-disable-this-line

      if (match) {
        let innerMatch: RegExpMatchArray | null;
        const nextLineRange = new OIRange(this.file, new OI(this.file, lineNumber + 1, 0), new OI(this.file, lineNumber + 1, this.originalText.split('\n')[lineNumber + 1].length - 1));
        if ((innerMatch = match[2].match('-disable-this-line')) !== null) {
          this.file.magicComments.push(new OMagicCommentDisable(this.file, MagicCommentType.Disable, new OIRange(this.file, new OI(this.file, lineNumber, 0), new OI(this.file, lineNumber, line.length - 1))));
        } else if ((innerMatch = match[2].match('-disable-next-line')) !== null) {// TODO: next nonempty line
          this.file.magicComments.push(new OMagicCommentDisable(this.file, MagicCommentType.Disable, nextLineRange));
        } else if ((innerMatch = match[2].match('-disable-region')) !== null) {
          if (disabledRangeStart === undefined) {
            disabledRangeStart = lineNumber;
          }
        } else if ((innerMatch = match[2].match('-enable-region')) !== null) {
          if (disabledRangeStart !== undefined) {
            const disabledRange = new OIRange(this.file, new OI(this.file, disabledRangeStart, 0), new OI(this.file, lineNumber, line.length - 1));
            this.file.magicComments.push(new OMagicCommentDisable(this.file, MagicCommentType.Disable, disabledRange));
            disabledRangeStart = undefined;
          }
        } else if ((innerMatch = match[2].match(/(-parameter-next-line\s+)(.*)/)) !== null) {// TODO: next nonempty line
          const parameter = innerMatch[2].split(/,?\s+/);
          // .map(parameter => {
          //   const innerInnerMatch = new RegExp(String.raw`\b${escapeStringRegexp(parameter)}\b`, 'i').exec((innerMatch as RegExpMatchArray)[2]);
          //   console.log(String.raw`\b${escapeStringRegexp(parameter)}\b`);
          //   if (!innerInnerMatch) {
          //     throw new Error('FUCK');
          //   }
          //   const startCharacter = (match as RegExpExecArray)[1].length + (innerMatch as RegExpMatchArray)[1].length + innerInnerMatch.index;
          //   const read = new ORead(this.file, new OI(this.file, lineNumber, startCharacter).i, new OI(this.file, lineNumber, startCharacter + innerInnerMatch.length - 1).i);
          //   read.text = parameter;
          //   return read;
          // });
          this.file.magicComments.push(new OMagicCommentParameter(this.file, MagicCommentType.Parameter, nextLineRange, parameter));
        } else if ((innerMatch = match[2].match(/-ignore\s+\/([^/]*)\/(.)?/)) !== null) {
          ignoreRegex.push(RegExp(innerMatch[1], innerMatch[2]));
        }
      }
      if (disabledRangeStart !== undefined) {
        const disabledRange = new OIRange(this.file, new OI(this.file, disabledRangeStart, 0), new OI(this.file, this.originalText.length - 1));
        this.file.magicComments.push(new OMagicCommentDisable(this.file, MagicCommentType.Disable, disabledRange));
        disabledRangeStart = undefined;
      }

    }
    for (const regex of ignoreRegex) {
      const ignores = this.text.match(regex);
      if (ignores === null) continue;
      for (const ignore of ignores) {
        const replacement = ignore.replace(/\S/g, ' ');
        this.text = this.text.replace(ignore, replacement);
      }
    }
    // this.pos = new OI(this.file, 0);
    if (this.text.length > 500 * 1024) {
      throw new ParserError('this.file too large', this.pos.getRangeToEndLine());
    }
    let contextReferences = [];
    const defaultLibrary = [
      new OLexerToken('std', new OIRange(this.file, 0, 0), TokenType.keyword),
      new OLexerToken('work', new OIRange(this.file, 0, 0), TokenType.keyword),
    ];
    const defaultUseClause = [
      new OUseClause(this.file, new OIRange(this.file, 0, 0), new OLexerToken('std', new OIRange(this.file, 0, 0), TokenType.keyword),
        new OLexerToken('standard', new OIRange(this.file, 0, 0), TokenType.keyword),
        new OLexerToken('all', new OIRange(this.file, 0, 0), TokenType.keyword))
    ];
    // TODO: Add library STD, WORK; use STD.STANDARD.all;
    let libraries = defaultLibrary.slice(0);
    let useClauses = defaultUseClause.slice(0);

    while (this.pos.isValid()) {
      this.advanceWhitespace();
      const nextToken = this.consumeToken();
      if (nextToken.getLText() === 'context') {
        if (this.advanceSemicolonToken(true, { consume: false }).find(token => token.getLText() === 'is')) {
          const contextParser = new ContextParser(this.pos, this.filePath, this.file);
          const context = contextParser.parse();
          context.libraries.push(...libraries.map(library => new OLibrary(context, library)));
          context.useClauses.push(...useClauses);
          context.contextReferences.push(...contextReferences);
          for (const contextReference of contextReferences) {
            contextReference.parent = context;
          }
          for (const useClause of useClauses) {
            useClause.parent = context;
          }

          libraries = defaultLibrary.slice(0);
          useClauses = defaultUseClause.slice(0);
          contextReferences = [];
          this.file.contexts.push(context);
        } else {
          // The Parent gets overwritten when attaching the reference to the correct object
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          const contextReferenceParser = new ContextReferenceParser(this.pos, this.filePath, (this.file as any));
          contextReferences.push(contextReferenceParser.parse());
        }
      } else if (nextToken.getLText() === 'library') {
        libraries.push(this.consumeToken());
        this.expect(';');
      } else if (nextToken.getLText() === 'use') {
        const useClauseParser = new UseClauseParser(this.pos, this.filePath, this.file);
        useClauses.push(useClauseParser.parse());
      } else if (nextToken.getLText() === 'entity') {
        const entityParser = new EntityParser(this.pos, this.filePath, this.file);
        const entity = entityParser.parse();
        this.file.entities.push(entity);
        entity.contextReferences = contextReferences;
        for (const contextReference of contextReferences) {
          contextReference.parent = entity;
        }
        for (const useClause of useClauses) {
          useClause.parent = entity;
        }
        contextReferences = [];
        entity.useClauses.push(...useClauses);
        entity.libraries.push(...libraries.map(library => new OLibrary(entity, library)));
        libraries = defaultLibrary.slice(0);
        useClauses = defaultUseClause.slice(0);
        if (this.onlyEntity) {
          break;
        }
        //         // console.log(this.file, typeof this.file.entity, 'typeof');
      } else if (nextToken.getLText() === 'architecture') {
        const architectureParser = new ArchitectureParser(this.pos, this.filePath, this.file);
        const architecture = architectureParser.parse();
        this.file.architectures.push(architecture);
        architecture.contextReferences = contextReferences;
        for (const contextReference of contextReferences) {
          contextReference.parent = architecture;
        }
        for (const useClause of useClauses) {
          useClause.parent = architecture;
        }
        contextReferences = [];
        architecture.useClauses.push(...useClauses);
        architecture.libraries.push(...libraries.map(library => new OLibrary(architecture, library)));

        libraries = defaultLibrary.slice(0);
        useClauses = defaultUseClause.slice(0);
      } else if (nextToken.getLText() === 'package') {
        if (this.onlyEntity && this.getNextWord({ consume: false }) === 'body') {
          // break;
        }
        const pkg = (this.getToken(2, true).getLText() === 'new')
          ? new PackageInstantiationParser(this.pos, this.filePath, this.file).parse()
          : new PackageParser(this.pos, this.filePath).parse(this.file);
        pkg.useClauses.push(...useClauses);
        pkg.libraries.push(...libraries.map(library => new OLibrary(pkg, library)));
        pkg.contextReferences = contextReferences;
        for (const contextReference of contextReferences) {
          contextReference.parent = pkg;
        }
        for (const useClause of useClauses) {
          useClause.parent = pkg;
        }
        if (pkg instanceof OPackageInstantiation) {
          this.file.packageInstantiations.push(pkg);
          this.expect(';'); // package instantiations do not parse ';'
        } else {
          this.file.packages.push(pkg);
        }
        contextReferences = [];
        libraries = defaultLibrary.slice(0);
        useClauses = defaultUseClause.slice(0);
      } else if (nextToken.getLText() === 'configuration') {
        const configuration = new OConfiguration(this.file, this.getToken().range.copyExtendEndOfLine());
        configuration.identifier = this.consumeToken();
        this.expect('of');
        configuration.entityName = this.consumeToken();
        while (
          ((this.getToken(0).getLText() === 'end' && this.getToken(1, true).getLText() === ';')
            || (this.getToken(0).getLText() === 'end' && this.getToken(1, true).getLText() === 'configuration'
              && this.getToken(2, true).getLText() === ';')
            || (this.getToken(0).getLText() === 'end' && this.getToken(1, true).getLText() === 'configuration'
              && this.getToken(2, true).getLText() === configuration.identifier.getLText() && this.getToken(3, true).getLText() === ';')
            || (this.getToken(0).getLText() === 'end'
              && this.getToken(1, true).getLText() === configuration.identifier.getLText() && this.getToken(2, true).getLText() === ';'))
          === false) {
          this.consumeToken(true);
        }
        this.file.configurations.push(configuration);
        this.advanceSemicolon();
      } else {
        throw new ParserError(`Unexpected token ${nextToken.text}`, this.getToken().range);
      }
    }
    return this.file;
  }
}
