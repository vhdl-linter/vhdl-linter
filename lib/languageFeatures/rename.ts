import { ErrorCodes, Position, ResponseError, TextEdit } from 'vscode-languageserver';
import { VhdlLinter } from "../vhdlLinter";
import { findReferenceAndDefinition, getTokenFromPosition } from "./findReferencesHandler";
import { OLexerToken } from '../lexer';
import { implementsIHasLabel } from '../parser/interfaces';

function isLabel(linter: VhdlLinter, token: OLexerToken) {
  return linter.file.objectList.some(obj => implementsIHasLabel(obj) && obj.label === token);
}

export function prepareRenameHandler(linter: VhdlLinter, position: Position) {

  const token = getTokenFromPosition(linter, position);
  if (!token || token.isIdentifier() === false || isLabel(linter, token)) {
    throw new ResponseError(ErrorCodes.InvalidRequest, 'Can not rename this element',  'Can not rename this element' );
  }

  return token.range;
}
export async function renameHandler(linter: VhdlLinter, position: Position, newName: string) {
  const tokens = await findReferenceAndDefinition(linter, position);
  const changes: Record<string, TextEdit[]> = {};
  for (const token of tokens) {
    const uri = token.file.uri.toString();
    const target = changes[uri] ?? [];
    if (!Array.isArray(changes[uri])) {
      changes[uri] = target;
    }
    target.push(TextEdit.replace(token.range, newName));
  }
  return {
    changes
  };
}