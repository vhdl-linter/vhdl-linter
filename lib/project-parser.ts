import { FSWatcher, watch } from 'chokidar';
import { EventEmitter } from 'events';
import { promises } from 'fs';
import { join, sep } from 'path';
import { cwd } from 'process';
import { OContext, OEntity, OPackage, OPackageInstantiation } from './parser/objects';
import { SettingsGetter, VhdlLinter } from './vhdl-linter';

export class ProjectParser {

  public cachedFiles: FileCache[] = [];
  public packages: OPackage[] = [];
  public packageInstantiations: OPackageInstantiation[] = [];
  public contexts: OContext[] = [];
  public entities: OEntity[] = [];
  events = new EventEmitter();
  private watchers: FSWatcher[] = [];
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(workspaces: string[], fileIgnoreRegex: string, settingsGetter: SettingsGetter, disableWatching = false) {
    const projectParser = new ProjectParser(workspaces, fileIgnoreRegex, settingsGetter);
    await projectParser.init(disableWatching);
    return projectParser;
  }
  private constructor(public workspaces: string[], public fileIgnoreRegex: string, public settingsGetter: SettingsGetter) { }
  public addFolders(pathes: string[]) {
    for (const path of pathes) {
      const watcher = watch(path.replace(sep, '/') + '/**/*.vhd?(l)', { ignoreInitial: true });
      watcher.on('add', async (path) => {
        const cachedFile = await (FileCache.create(path, this));
        this.cachedFiles.push(cachedFile);
        this.flattenProject();
        this.events.emit('change', 'add', path);
      });
      watcher.on('change', async (path) => {
        // console.log('change', path);
        const cachedFile = process.platform === 'win32'
          ? this.cachedFiles.find(cachedFile => cachedFile.path.toLowerCase() === path.toLowerCase())
          : this.cachedFiles.find(cachedFile => cachedFile.path === path);
        if (cachedFile) {
          await cachedFile.parse();
          this.flattenProject();
          this.events.emit('change', 'change', path);
        } else {
          console.error('modified file not found', path);
        }
      });
      watcher.on('unlink', path => {
        const cachedFileIndex = this.cachedFiles.findIndex(cachedFile => cachedFile.path === path);
        if (cachedFileIndex > -1) {
          this.cachedFiles.splice(cachedFileIndex, 1);
          this.flattenProject();
          this.events.emit('change', 'unlink', path);
        }
      });
      this.watchers.push(watcher);
    }
  }
  private async init(disableWatching: boolean) {
    const files = new Set<string>();
    await Promise.all(this.workspaces.map(async (directory) => {
      const directories = await this.parseDirectory(directory);
      return (await Promise.all(directories.map(file => promises.realpath(file)))).forEach(file => files.add(file));
    }));
    (await this.parseDirectory(join(cwd(), 'ieee2008'))).forEach(file => files.add(file));

    for (const file of files) {
      const cachedFile = await FileCache.create(file, this);
      this.cachedFiles.push(cachedFile);
    }
    this.cachedFiles.sort((a, b) => b.lintingTime - a.lintingTime);
    // console.log('Times: \n' + this.cachedFiles.slice(0, 10).map(file => `${file.path}: ${file.lintingTime}ms`).join('\n'));
    this.flattenProject();
    if (!disableWatching) {
      this.addFolders(this.workspaces);
    }
  }
  async stop() {
    for (const watcher of this.watchers) {
      await watcher.close();
    }
  }
  private async parseDirectory(directory: string): Promise<string[]> {
    const files: string[] = [];
    const entries = await promises.readdir(directory);
    const ignoreRegex = this.fileIgnoreRegex.trim().length > 0 ? new RegExp(this.fileIgnoreRegex) : null;

    // const entries = await promisify(directory.getEntries)()
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = directory + sep + entry;
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile()) {
          if (entry.match(/\.vhdl?$/i) && (ignoreRegex === null || !filePath.match(ignoreRegex))) {
            files.push(filePath);
          }
        } else {
          files.push(... await this.parseDirectory(filePath));
        }
      } catch (e) {
        console.error(e);
      }
    }));
    return files;
  }
  private flattenProject() {
    this.packages = [];
    this.packageInstantiations = [];
    this.entities = [];
    for (const cachedFile of this.cachedFiles) {
      this.entities.push(...cachedFile.entities);
      if (cachedFile.packages) {
        this.packages.push(...cachedFile.packages);
      }
      if (cachedFile.packageInstantiations) {
        this.packageInstantiations.push(...cachedFile.packageInstantiations);
      }
      if (cachedFile.contexts) {
        this.contexts.push(...cachedFile.contexts);
      }
    }
  }

}

export class FileCache {
  packages?: OPackage[];
  packageInstantiations?: OPackageInstantiation[];
  contexts: OContext[] = [];
  entities: OEntity[] = [];
  text: string;
  linter: VhdlLinter;
  lintingTime: number;
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(path: string, projectParser: ProjectParser) {
    const cache = new FileCache(path, projectParser);
    await cache.parse();
    return cache;
  }
  private constructor(public path: string, public projectParser: ProjectParser) {
  }
  async parse() {
    this.text = await promises.readFile(this.path, { encoding: 'utf8' });
    this.linter = new VhdlLinter(this.path, this.text, this.projectParser, this.projectParser.settingsGetter);
    this.packages = this.linter.file.packages.filter((p): p is OPackage => p instanceof OPackage);
    this.packageInstantiations = this.linter.file.packageInstantiations;
    this.entities = this.linter.file.entities;
    this.contexts = this.linter.file.contexts;
  }

}