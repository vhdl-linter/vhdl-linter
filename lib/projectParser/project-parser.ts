import { FSWatcher, watch } from 'chokidar';
import { EventEmitter } from 'events';
import { promises, readFileSync } from 'fs';
import { Worker } from 'node:worker_threads';
import { join, sep } from 'path';
import { OContext, OEntity, OFile, OPackage, OPackageBody } from '../parser/objects';
import * as objects from '../parser/objects';
import { VhdlLinter } from '../vhdl-linter';
// import { StaticPool } from 'node-worker-threads-pool';
import { Parser } from '../parser/parser';
import { pool } from './parser-pool';
export class ProjectParser {

  public cachedFiles: OFileCache[] = [];
  private packages: (OPackage | OPackageBody)[];
  private contexts: OContext[] = [];
  private entities: OEntity[];
  events = new EventEmitter();
  constructor(public workspaces: string[], public fileIgnoreRegex: string) { }
  async init() {
    const files = new Set<string>();
    await Promise.all(this.workspaces.map(async (directory) => {
      // console.log('dir', directory);
      const directories = await this.parseDirectory(directory);
      return (await Promise.all(directories.map(file => promises.realpath(file)))).forEach(file => files.add(file));
    }));
    // for (const directory of this.workspaces) {
    //   this.parseDirectory(directory).forEach(file => files.add(realpathSync(file)));
    // }
    const pkg = __dirname;
    if (pkg) {
      //       console.log(pkg, new Directory(pkg + '/ieee2008'));
      (await this.parseDirectory(join(pkg, `${sep}..${sep}..${sep}..${sep}ieee2008`))).forEach(file => files.add(file));
    }
    for (const file of files) {
      const cachedFile = new OFileCache(file);
      await cachedFile.parse();
      this.cachedFiles.push(cachedFile);
    }
    // this.cachedFiles.push(...await Promise.all(Array.from(files).map(async file => {
    //   const cachedFile = new OFileCache(file, this);
    //   await cachedFile.parse();
    //   return cachedFile;
    // })));
    this.cachedFiles.sort((a, b) => b.lintingTime - a.lintingTime);
    // console.log('Times: \n' + this.cachedFiles.slice(0, 10).map(file => `${file.path}: ${file.lintingTime}ms`).join('\n'));
    this.fetchEntitesAndPackagesAndContexts();
    for (const workspace of this.workspaces) {
      const watcher = watch(workspace.replace(sep, '/') + '/**/*.vhd', { ignoreInitial: true });
      watcher.on('add', async (path) => {
        const cachedFile = new OFileCache(path);
        await cachedFile.parse();
        this.cachedFiles.push(cachedFile);
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'add', path);
      });
      watcher.on('change', async (path) => {
        // console.log('change', path);
        const cachedFile = process.platform === 'win32'
          ? this.cachedFiles.find(cachedFile => cachedFile.path.toLowerCase() === path.toLowerCase())
          : this.cachedFiles.find(cachedFile => cachedFile.path === path);
        if (cachedFile) {
          await cachedFile.parse();
        } else {
          console.error('modified file not found', path);
        }
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'change', path);
      });
      watcher.on('unlink', path => {
        const cachedFileIndex = this.cachedFiles.findIndex(cachedFile => cachedFile.path === path);
        this.cachedFiles.splice(cachedFileIndex, 1);
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'unlink', path);
      });

    }
  }
  private async parseDirectory(directory: string): Promise<string[]> {
    const files: string[] = [];
    const entries = await promises.readdir(directory);
    const ignoreRegex = this.fileIgnoreRegex.trim().length > 0 ? new RegExp(this.fileIgnoreRegex) : null;

    // const entries = await promisify(directory.getEntries)()
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = directory + sep + entry;
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile()) {
          if (entry.match(/\.vhdl?$/i) && (ignoreRegex === null || !filePath.match(ignoreRegex))) {
            files.push(filePath);
          }
        } else {
          files.push(... await this.parseDirectory(filePath));
        }
      } catch (e) {
        console.error(e);
      }
    }));
    return files;
  }
  private fetchEntitesAndPackagesAndContexts() {
    //     console.log(this.cachedFiles);
    this.packages = [];
    this.entities = [];
    for (const cachedFile of this.cachedFiles) {
      this.entities.push(...cachedFile.entities);
      if (cachedFile.packages) {
        this.packages.push(...cachedFile.packages);
      }
      if (cachedFile.contexts) {
        this.contexts.push(...cachedFile.contexts);
      }
    }
  }
  watcher: FSWatcher;
  addFolders(folders: string[]) {
    this.watcher.add(folders.map(folder => folder.replace(sep, '/') + '/**/*.vhd'));
  }
  public getPackages() {
    return this.packages;
  }
  public getContexts() {
    return this.contexts;
  }
  public getEntities() {
    return this.entities;
  }
}

export class OFileCache {

  path: string;
  digest: string;
  file: OFile;
  packages?: (OPackage | OPackageBody)[];
  contexts: OContext[] = [];
  entities: OEntity[] = [];
  lintingTime: number;
  constructor(file: string) {
    // this.digest = await file.getDigest();
    this.path = file;
  }
  async parse() {
    console.log(this.path, 'start');
    // const parser = new Parser(readFileSync(this.path, { encoding: 'utf-8' }), this.path, true);
    // this.file = parser.parse();

    this.file = await pool.exec({
      path: this.path,
    });
    console.log(this.path, 'end');
    this.packages = this.file.packages;
    this.entities = this.file.entities;
    this.contexts = this.file.contexts;
  }
}