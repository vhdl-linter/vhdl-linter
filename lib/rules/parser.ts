import { RuleBase, IRule } from "./rules-base";
import { OFile } from "../parser/objects";

export class RParser extends RuleBase implements IRule {
  public name = 'parser';
  file: OFile;

  async check() {
    for (const message of this.file.parserMessages) {
      this.addMessage(message);
    }
  }
}