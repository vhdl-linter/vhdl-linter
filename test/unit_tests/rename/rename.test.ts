import { expect, test, beforeAll, afterAll } from '@jest/globals';
import { readFileSync } from 'fs';
import { ErrorCodes, Position, ResponseError } from 'vscode-languageserver';
import { prepareRenameHandler, renameHandler } from '../../../lib/languageFeatures/rename';
import { OIRange } from '../../../lib/parser/objects';
import { ProjectParser } from '../../../lib/project-parser';
import { defaultSettingsGetter } from '../../../lib/settings';
import { VhdlLinter } from '../../../lib/vhdl-linter';
import { createPrintableRange, makeRangePrintable } from '../../helper';
let projectParser: ProjectParser;
beforeAll(async () => {
  projectParser = await ProjectParser.create([__dirname], '', defaultSettingsGetter);
});
afterAll(async () => {
  await projectParser.stop();
});
// use dummy Path to avoid having path in snapshots

test.each([
  // testing stuff in one file
  ['entity.vhd', createPrintableRange(5, 8, 5, 19), 3], // three occurrences of entity name (test_entity)
  ['entity.vhd', createPrintableRange(6, 5, 6, 16), 3],
  ['entity.vhd', createPrintableRange(8, 22, 8, 33), 3],
  ['entity.vhd', createPrintableRange(5, 1, 5, 7), 0], // No rename entity keyword
  ['entity.vhd', createPrintableRange(8, 1, 8, 13), 0], // No rename architecture keyword
  ['entity.vhd', createPrintableRange(8, 14, 8, 18), 2], // two occurrences of architecture name (arch)
  ['entity.vhd', createPrintableRange(11, 18, 11, 22), 2],
  ['signal.vhd', createPrintableRange(11, 10, 11, 13), 3], // three occurrences of the signal name foo
  ['signal.vhd', createPrintableRange(16, 3, 16, 6), 3],
  ['signal.vhd', createPrintableRange(17, 24, 17, 27), 3],

  // testing stuff over multiple files
  ['signal.vhd', createPrintableRange(13, 21, 13, 27), 3], // 3 occurences of t_enum
  ['package.vhd', createPrintableRange(5, 8, 5, 14), 3],
  ['package.vhd', createPrintableRange(9, 15, 9, 21), 3],

  ['signal.vhd', createPrintableRange(14, 21, 14, 29), 2], // 2 occurences of t_record
  ['package.vhd', createPrintableRange(7, 8, 7, 16), 2],

  ['signal.vhd', createPrintableRange(19, 12, 19, 20), 2], // 2 occurences of element1 of t_record
  ['package.vhd', createPrintableRange(9, 5, 9, 13), 2],

  ['package.vhd', createPrintableRange(4, 9, 4, 17), 5], // 5 occurrences of package name (test_pkg)
  ['signal.vhd', createPrintableRange(4, 10, 4, 18), 5], // also expect renaming in signal.vhd:4
  ['package.vhd', createPrintableRange(12, 13, 12, 21), 5],
  ['package.vhd', createPrintableRange(13, 14, 13, 22), 5],
  ['package.vhd', createPrintableRange(14, 18, 14, 26), 5],
  // Testing split entity and architecture file
  ['entity_split.vhd', createPrintableRange(2, 8, 2, 25), 3], // two occurrences in entity file
  ['entity_split.vhd', createPrintableRange(3, 5, 3, 22), 3],
  ['architecture_split.vhd', createPrintableRange(1, 22, 1, 39), 3],

])('testing rename for %s in %s (expected %d occurences)', async (name, range, expectOccurences) => {
  const path = __dirname + `/${name}`;
  const dummyDirname = `/file`;

  const linter = new VhdlLinter(path, readFileSync(path, { encoding: 'utf8' }),
    projectParser, defaultSettingsGetter);
  await linter.checkAll();
  await projectParser.stop();
  const newName = 'foo_bar';
  const pos = range.start;
  const { start, end } = range;
  if (expectOccurences > 0) {
    let lastOperations;
    for (let character = start.character; character <= end.character; character++) {
      const result = await prepareRenameHandler(linter, Position.create(start.line, character));
      expect(result).toBeInstanceOf(OIRange);
      expect(result.start.line).toBe(start.line);
      expect(result.start.character).toBe(start.character);
      expect(result.end.line).toBe(end.line);
      expect(result.end.character).toBe(end.character);
      const renameOperations = await renameHandler(linter, pos, newName);
      const operations = Object.entries(renameOperations.changes).map(([file, ops]) => ({ file: file.replace(__dirname, dummyDirname), changes: ops.map(op => makeRangePrintable(op.range)) }));
      expect(operations.flatMap(o => o.changes)).toHaveLength(expectOccurences);
      if (lastOperations) {
        expect(operations).toStrictEqual(lastOperations);
      } else {
        expect(operations).toMatchSnapshot();
        lastOperations = operations;
      }

    }
  } else {
    for (let character = start.character; character <= end.character; character++) {

      let err;
      try {
        await prepareRenameHandler(linter, Position.create(start.line, character));
      } catch (_err) {
        err = _err;
      }
      expect(err).toBeInstanceOf(ResponseError);
      expect(err.message).toBe('Can not rename this element');
    }
  }
});
test('testing handling of invalid rename Handler', async () => {
  const filename = 'entity.vhd';
  const path = __dirname + `/${filename}`;
  const dummyPath = `/file/${filename}`;
  const linter = new VhdlLinter(dummyPath, readFileSync(path, { encoding: 'utf8' }),
    projectParser, defaultSettingsGetter);
  await linter.checkAll();
  await projectParser.stop();
  let err;
  try {
    await renameHandler(linter, Position.create(7, 1), 'test');
  } catch (_err) {
    err = _err;
  }
  expect(err).toBeInstanceOf(ResponseError);
  expect(err.message).toBe('Can not rename this element');
  expect(err.code).toBe(ErrorCodes.InvalidRequest);
});