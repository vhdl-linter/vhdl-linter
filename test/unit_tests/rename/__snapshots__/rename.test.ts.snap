// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing rename for entity.vhd in 5:8 - 5:19 true 1`] = `
[
  "5:8 - 5:19",
  "8:22 - 8:33",
  "6:5 - 6:16",
]
`;

exports[`testing rename for entity.vhd in 6:5 - 6:16 true 1`] = `
[
  "5:8 - 5:19",
  "8:22 - 8:33",
  "6:5 - 6:16",
]
`;

exports[`testing rename for entity.vhd in 8:14 - 8:18 true 1`] = `
[
  "8:14 - 8:18",
  "11:18 - 11:22",
]
`;

exports[`testing rename for entity.vhd in 8:22 - 8:33 true 1`] = `
[
  "5:8 - 5:19",
  "8:22 - 8:33",
  "6:5 - 6:16",
]
`;

exports[`testing rename for entity.vhd in 11:18 - 11:22 true 1`] = `
[
  "8:14 - 8:18",
  "11:18 - 11:22",
]
`;

exports[`testing rename for package.vhd in 1:9 - 1:17 true 1`] = `
[
  "1:9 - 1:17",
  "3:13 - 3:21",
  "4:14 - 4:22",
  "5:18 - 5:26",
]
`;

exports[`testing rename for package.vhd in 3:13 - 3:21 true 1`] = `
[
  "1:9 - 1:17",
  "3:13 - 3:21",
  "4:14 - 4:22",
  "5:18 - 5:26",
]
`;

exports[`testing rename for package.vhd in 4:14 - 4:22 true 1`] = `
[
  "1:9 - 1:17",
  "3:13 - 3:21",
  "4:14 - 4:22",
  "5:18 - 5:26",
]
`;

exports[`testing rename for package.vhd in 5:18 - 5:26 true 1`] = `
[
  "1:9 - 1:17",
  "3:13 - 3:21",
  "4:14 - 4:22",
  "5:18 - 5:26",
]
`;

exports[`testing rename for signal.vhd in 10:10 - 10:13 true 1`] = `
[
  "10:10 - 10:13",
  "13:3 - 13:6",
  "14:10 - 14:13",
]
`;

exports[`testing rename for signal.vhd in 13:3 - 13:6 true 1`] = `
[
  "10:10 - 10:13",
  "13:3 - 13:6",
  "14:10 - 14:13",
]
`;
